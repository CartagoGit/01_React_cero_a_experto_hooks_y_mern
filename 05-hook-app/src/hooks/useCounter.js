import { useState } from "react";

export const useCounter = (initialState = 0) => {
	const [counter, setCounter] = useState(initialState);

	const increment = (num=1) => {
		setCounter(counter => counter + num);
	};
	const decrement = (num=1) => {
		setCounter(counter => counter - num);
	};
	const reset = () => {
		setCounter(counter => counter = initialState);
	};

	return {
		counter,
		increment,
		decrement,
		reset,
		setCounter
	};
};
