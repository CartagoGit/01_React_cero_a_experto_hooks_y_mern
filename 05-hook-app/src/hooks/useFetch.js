import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
	const isMounted = useRef(true);
	const [state, setState] = useState({
		data: null,
		loading: true,
		error: null
	});

	useEffect(() => {
		return () => {
			isMounted.current = false;
		};
	}, []);

	useEffect(() => {
		setState({ data: null, loading: true, error: null });
		fetch(url)
			.then((resp) => resp.json())
			.then((data) => {
				// aqui validamos el arreglo no este en cero
				// console.log(url);
				// if (data.length === 0)
				// 	data.push({
				// 		quote: "Without Data"
				// 	});
				if (data.length === 0) throw new EmptyStackException();
				if (isMounted.current) {
					setState({
						loading: false,
						error: null,
						data
					});
				}
			})
			.catch((error) =>
				setState({
					loading: false,
					error: 'Error fetching url',
					data: null
				})
			);
	}, [url]);

	return state;
};
