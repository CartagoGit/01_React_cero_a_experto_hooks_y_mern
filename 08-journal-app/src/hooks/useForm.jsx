import { useState } from "react";

export const useForm = (initialState = {}) => {
	const [values, setValues] = useState(initialState);	const reset = (newFormState = initialState) => {
		setValues(() => newFormState);
	};

	const handleChangeValue = ({ target }) => {
		setValues({
			...values,
			[target.name]: target.value
		});
	};

	return [values, handleChangeValue, reset];
};
