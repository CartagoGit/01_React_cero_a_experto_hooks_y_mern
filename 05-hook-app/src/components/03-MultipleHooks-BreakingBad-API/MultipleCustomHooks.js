import React from "react";
import { useCounter } from "../../hooks/useCounter";
import { useFetch } from "../../hooks/useFetch";
import "../02-useEffect/effects.css";

export const MultipleCustomHooks = () => {
	const { counter, increment, decrement, setCounter } = useCounter(30);
	const maxCounter = [30, 102];
	const minCounter = [1, 63];
	const { loading, data } = useFetch(
		`https://www.breakingbadapi.com/api/quotes/${counter}`
	);
	const { author, quote = "There are not quote with that Id" } =
		!!data && !!data[0] && data[0];

	const next = () => {
		increment();
		counter === maxCounter[0] && setCounter(minCounter[1]);
	};

	const last = () => {
		decrement();
		counter === minCounter[1] && setCounter(maxCounter[0]);
	};

	return (
		<div>
			<h1>BreakingBad's Quotes</h1>
			<hr />
			{loading ? (
				<div className='alert alert-info text-center'>Loading...</div>
			) : (
				<blockquote className='blockquote text-end'>
					<p className='mb-3'>{quote}</p>
					<footer className='blockquote-footer'>{author}</footer>
				</blockquote>
			)}
            <hr/>
			<div className='d-flex justify-content-between'>
				{minCounter[0] < counter && (
					<button onClick={last} className='btn btn-primary'>
						Last Quote
					</button>
				)}

				{counter < maxCounter[1] && (
					<button onClick={next} className='btn btn-primary ms-auto'>
						Next Quote
					</button>
				)}
			</div>
		</div>
	);
};
