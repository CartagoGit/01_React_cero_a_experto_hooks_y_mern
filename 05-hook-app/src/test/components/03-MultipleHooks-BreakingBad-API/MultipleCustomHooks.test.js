import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-MultipleHooks-BreakingBad-API/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("Pruebas en <MultipleCustomHooks />", () => {
	let wrapper;

	const fillWrapper = () => {
		useCounter.mockReturnValue({
			counter: 10
			// increment: () => {},
			// decrement: () => {}
		});
		useFetch.mockReturnValue({ data: null, loading: true, error: null }); //al usar el mock para que no compruebe el useFetch
		wrapper = shallow(<MultipleCustomHooks />);
	};

	beforeEach(() => fillWrapper());

	test("Deberia mostrarse correctamente", () => {
		expect(wrapper).toMatchSnapshot();
	});

	test("Deberia mostrar la informaciÃ³n", () => {
		useFetch.mockReturnValue({
			data: [{ author: "Mario", quote: "Hola Mundo" }],
			loading: false,
			error: null
		});
		wrapper = shallow(<MultipleCustomHooks />);

		expect(wrapper.find(".alert").exists()).toBe(false);
		expect(wrapper.find("p").text().trim()).toBe("Hola Mundo");
		expect(wrapper.find("footer").text().trim()).toBe("Mario");
	});
});
