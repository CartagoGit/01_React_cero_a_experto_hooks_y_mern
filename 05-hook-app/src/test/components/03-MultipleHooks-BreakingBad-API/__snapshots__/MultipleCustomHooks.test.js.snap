// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Deberia mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBad's Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <hr />
  <div
    className="d-flex justify-content-between"
  >
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Last Quote
    </button>
    <button
      className="btn btn-primary ms-auto"
      onClick={[Function]}
    >
      Next Quote
    </button>
  </div>
</div>
`;
